include/tscpp/api/HttpMsgComp.h

```diff
@@ -0,0 +1,424 @@
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+
+/**
+ * @file HttpMsgComp.h
+ *
+ * Classes for convenient manipulation of HTTP message components.
+ */
+
+/*
+Note:  The clases in this include file may be used independently from the rest of the C++ API.
+*/
+
+#pragma once
+
+// Suppress assert() macro in non-debug builds.
+#if !defined(NDEBUG) && !defined(DEBUG)
+#define NDEBUG
+#endif
+#include <cassert>
+
+#include <string_view>
+
+#include <ts/ts.h>
+
+#include <tscpp/api/noncopyable.h>
+
+/*
+Note:  These clases are designed to create intances as local variables in functions.  It is expected that compiler
+optimization will prevent the use of references from creating run-time overhead (see https://godbolt.org/z/KJZkno ).
+*/
+
+namespace atscppapi
+{
+class MsgBuffer : noncopyable
+{
+public:
+  MsgBuffer(TSMBuffer hndl) : _hndl(hndl) {}
+
+  // Returns the handle.
+  //
+  TSMBuffer
+  bufHndl() const
+  {
+    return _hndl;
+  }
+
+protected:
+  MsgBuffer() {}
+
+  void
+  set(TSMBuffer hndl)
+  {
+    _hndl = hndl;
+  }
+
+private:
+  TSMBuffer _hndl;
+};
+
+class MsgBase;
+
+class MimeField
+{
+public:
+  explicit MimeField(MsgBase &msg, TSMLoc loc = TS_NULL_MLOC) : _msg(msg), _loc(loc) {}
+
+  // MimeField at (zero-base) index idx in HTTP message.
+  //
+  MimeField(MsgBase &msg, int idx);
+
+  // MimeField with given name in HTTP message.
+  //
+  MimeField(MsgBase &msg, std::string_view name);
+
+  TSMLoc
+  loc() const
+  {
+    return _loc;
+  }
+
+  bool valid() const;
+
+  void reset();
+
+  // Can not copy, only move.
+
+  MimeField(const MimeField &) = delete;
+  MimeField &operator=(const MimeField &) = delete;
+
+  MimeField(MimeField &&source);
+
+  // Move assign only allowed if the two instances refer to the same message.
+  //
+  MimeField &operator=(MimeField &&source);
+
+  void destroy();
+
+  // Next field, returns invalid instance if none.
+  //
+  MimeField next() const;
+
+  // Next field with same name, return invalid instance if none.
+  //
+  MimeField nextDup() const;
+
+  std::string_view nameGet() const;
+
+  void nameSet(std::string_view);
+
+  void valuesClear();
+
+  // Returns number of values. Values index are from 0 to valuesCount() - 1.
+  //
+  int valuesCount() const;
+
+  std::string_view valGet(int idx) const;
+
+  void valSet(int idx, std::string_view new_value);
+
+  // Insert a new value at index idx.  If idx == valuesCount(), the new value is appended at the end.  Otherwise, all
+  // values with index >= idx prior to calling this have their index incremented by one.
+  //
+  void valInsert(int idx, std::string_view new_value);
+
+  ~MimeField();
+
+private:
+  MsgBase &_msg;
```

Comment:
I think it's better to copy over the `MBuffer` and `MLoc` instead of carrying a reference. It's cheap to do and removes any scope dependencies between them. It's quite possible to envision logic that gets the field and passes it on after the base object has gone out of scope.

Reply:
I can't envision logic where it would be desirable to have multiple copies of these.  Especially given the problem that they may go stale if you carry them from one hook to another.  I gave a link to an example in the comments showing that gcc is good at optimizing out references, even when they are class members.

Reply:
I think it's better to have copies than a hidden lifetime dependency. The copies are cheap. Going stale is the same issue in either case - I don't see why it matters if the staleness is via reference or copy. In both cases all instances need to have been cleaned up.

Consider code like
```
MimeField host;
if (some_flag) {
  HttpMsg msg = get_proxy_request();
  host = msg.findMimeField("Host");
} else {
  HttpMsg msg = get_proxy_response();
  host = msg.findMimeField("Host");
}
field.assign(pristine_host); // Woops, field depends on an out of scope object.
```

Reply:
Even with copies, your example is problematic.  My current design is that ~HttpMsg() releases the TMLoc for the message.  If it's not done in the destructor, then where?  Why not:

```
HttpMsg msgReq = get_proxy_request();
HttpMsg msgRes = get_proxy_response();
MimeField host = (some_flag ? msgReq : msgResp).findMimeField("Host");
host.assign(pristine_host);
```

Reply:
Each object tracks its own `TSMLoc` and releases it when destructed.

Your example could be simplified down to
```
MimeFile host = (some_flag ? get_proxy_request() : get_proxy_response()).findMimeField("Host");
host.assign(pristine_host);
```
Which causes the problem again. Certainly the reference style could be made to work, but it imposes additional very non-obvious requirements on how the classes can be used, for very little (if any) benefit.

Reply:
I guess what you're saying is that MimeField doesn't get its MBuffer and TSMLoc instances by a simple assign, it gets new values that refer to the same buffer and message header.  What is the TS API call to get these new values?  Is there a TS API call that can check that two distinct MBuffer/TSMLoc instances refer to the same object?

Certainly there are advantages to having pointers be shared_ptr or something similar (or using some other form of garbage collection).  But it's not the C++ way to use this approach exclusively.  Rightly or wrongly, the perception is that the CPP API is too high-overhead.  So that argues against increasing overhead to make it perfectly ironclad.

Reply:
It looks like you are making multiple simple copies of the TSMLoc for the message headers:  https://github.com/SolidWallOfCode/txn_box/blob/03093bf02e46f092691afe70d7b3d87151b2953a/plugin/include/txn_box/ts_util.h#L116 .  Seems like your are leaking them.eiddccgeijngvkttvu

Reply:
OK looking at TSHandleMLocRelease(), it doesn't actually do anything for message header mlocs (HDR_HEAP_OBJ_HTTP_HEADER), are you saying we should rely on that fact?

Reply:
For this function, https://github.com/SolidWallOfCode/txn_box/blob/03093bf02e46f092691afe70d7b3d87151b2953a/plugin/src/txn_box.cc#L122 , at what point does the returned TextView become stale?

Reply:
BTW (some_flag ? get_proxy_request() : get_proxy_response()).findMimeField("Host") would not compile in my stuff because moving is deleted for the message classes.